src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php

```diff
@@ -3429,43 +3429,52 @@ private static function getInarrayAssertions(
                     || $atomic_type instanceof Type\Atomic\TKeyedArray
                     || $atomic_type instanceof Type\Atomic\TList
                 ) {
+                    $is_sealed = false;
                     if ($atomic_type instanceof Type\Atomic\TList) {
                         $value_type = $atomic_type->type_param;
                     } elseif ($atomic_type instanceof Type\Atomic\TKeyedArray) {
                         $value_type = $atomic_type->getGenericValueType();
+                        $is_sealed = $atomic_type->sealed;
                     } else {
                         $value_type = $atomic_type->type_params[1];
                     }
 
-                    $array_literal_types = \array_filter(
-                        $value_type->getAtomicTypes(),
-                        function ($type) {
-                            return $type instanceof Type\Atomic\TLiteralInt
-                                || $type instanceof Type\Atomic\TLiteralString
-                                || $type instanceof Type\Atomic\TLiteralFloat
-                                || $type instanceof Type\Atomic\TEnumCase;
+                    $assertions = [];
+
+                    foreach ($value_type->getAtomicTypes() as $atomic_value_type) {
+                        $assertion = '';
+                        // If it's not a sealed (fixed, known) array, we can't simply return value types as assertions.
+                        // E. g. in_array($x, ['a', 'b']) is the same as $x === 'a' || $x === 'b',
+                        // which can also be negated correctly.
+                        // However, in_array($x, $y), where y is list<'a'|'b'> doesn't work the same way.
+                        // With positive assertion it has similar meaning: $x is 'a'|'b'.
+                        // But without knowing exact contents of $y, it's not an equality assertion
+                        // threfore it cannot be safely negated.
+                        // If we simply return =string(a) and =string(b) assertions, they will have the same semantics
+                        // as in the first example. So when negated, we will end up with $x !== 'a' && $x !== 'b'.
+                        // That won't work for unknown (not sealed) haystack.
+                        // 'in-array-' prefix is added to distinguish such assertions.
+                        if (!$is_sealed) {
+                            $assertion .= 'in-array-';
```

Comment:
Is it possible to fall in `!is-sealed` and in the first block of conditions below? `in-array-=...` would fail I think

If it's not possible, could we rearrange conditions to make that more clear?

Reply:
Well, technically such case is possible. Haystack of type `list<'a'|'b'|int>` and alike will cause that.
That's why there is a try-catch around `Type::parseString()` in `SimpleAssertionReconciler::reconcileInArray()`. The idea was to leave it as an option for future improvement if anyone ever wants to type-check such weird arrays.

I didn't originally intend to pursue it, but I ended up finding a bunch of cases, which didn't produce satisfactory results, like some unions of scalar types (e. g. `int|bool` needle with `array<int|string>` haystack etc.) triggering somewhat misleading/incorrect errors. So I went down this rabbit hole and in the end there's no more `in-array-=...` case.

I'll push the changes and post more details shortly.